shiny/render/_data_frame_utils/_selection.py

```diff
@@ -0,0 +1,450 @@
+    BrowserCellSelectionNone,
+]
+
+
+def assert_rows(x: BrowserCellSelection):
+    assert "rows" in x
+    assert isinstance(x["rows"], tuple) and len(x["rows"]) == 2
+    assert isinstance(x["rows"][0], int)
+    assert isinstance(x["rows"][1], int)
+
+
+def assert_cols(x: BrowserCellSelection):
+    assert "cols" in x
+    assert isinstance(x["cols"], tuple) and len(x["cols"]) == 2
+    assert isinstance(x["cols"][0], int)
+    assert isinstance(x["cols"][1], int)
+
+
+def is_location_range(r: LocationRange) -> TypeGuard[LocationRange]:
+    return (
+        isinstance(r, tuple)
+        and len(r) == 2
+        and isinstance(r[0], int)
+        and isinstance(r[1], int)
+    )
+
+
+def is_sequence_2(r: Sequence[int]) -> TypeGuard[LocationRange]:
+    return (
+        isinstance(r, tuple)
+        and len(r) == 2
+        and isinstance(r[0], int)
+        and isinstance(r[1], int)
+    )
+
+
+# def as_cell_selection(
+#     x: SelectionLocationJS | CellSelectionDict,
+# ) -> SelectionLocationJS:
+#     if x == "all":
+#         return "all"
+#     if x == "none":
+#         return "none"
+
+#     assert isinstance(x, dict)
+
+#     # Seems like a SelectionLocationJS, verify shape...
+#     if "type" in x:
+#         if x["type"] == "all":
+#             return x
+
+#         if x["type"] == "row":
+#             assert_rows(x)
+#             return x
+#         if x["type"] == "col":
+#             assert_cols(x)
+#             return x
+#         if x["type"] == "region":
+#             assert_rows(x)
+#             assert_cols(x)
+#             return x
+#         raise ValueError(f"Unknown SelectionLocationJS[\"type\"]: {x['type']}")
+
+#     # Seems like a SelectionLocation, verify and convert...
+#     if "rows" not in x:
+#         raise ValueError("SelectionLocation must have 'rows' key")
+#     if "cols" not in x:
+#         raise ValueError("SelectionLocation must have 'cols' key")
+
+#     # All
+#     if x["rows"] == "all" and x["cols"] == "all":
+#         return {"type": "all"}
+
+#     # Cols
+#     if x["rows"] == "all":
+#         assert x["cols"] != "all"
+#         if is_location_range(x["cols"]):
+#             return {"type": "col", "cols": x["cols"]}
+#         else:
+#             raise ValueError("SelectionLocation['cols'] must be a tuple[int, int]")
+
+#     # Rows
+#     if x["cols"] == "all":
+#         assert x["rows"] != "all"
+#         if is_location_range(x["rows"]):
+#             return {"type": "row", "rows": x["rows"]}
+#         else:
+#             raise ValueError("SelectionLocation['rows'] must be a tuple[int, int]")
+
+
+#     # Region
+#     if is_location_range(x["rows"]) and is_location_range(x["cols"]):
+#         return {"type": "region", "rows": x["rows"], "cols": x["cols"]}
+#     else:
+#         raise ValueError(
+#             "SelectionLocation['rows'] and ['cols'] must be a tuple[int, int]"
+#         )
+
+
+def to_list_or_none(x: int | Sequence[int] | None) -> Sequence[int] | None:
```

Comment:
```suggestion
def to_list_or_none(x: int | Sequence[int] | None) -> list[int] | None:
```

Also, I'm not sure if this will be helpful, but you could add overrides which say that `None` will go to `None`, and `int | Sequence[int]` will go to `list[int]`.

Reply:
The overload type did not work when it was being used. Skipping.

Suggestion: √